spec_util/meld.go

```diff
@@ -308,50 +308,37 @@ func dataEqual(dst, src *pb.Data) bool {
 	return proto.Equal(dst, src)
 }
 
-func recordConflict(dst, src *pb.Data) error {
-	// Special case: If and only if one data has a type hint, assign it to the other
-	// data so that the difference does not trigger a conflict and the type hint is preserved.
-	{
-		srcTypeHint := getTypeHint(src)
-		dstTypeHint := getTypeHint(dst)
-		if srcTypeHint != dstTypeHint {
-			if dstTypeHint == "" {
-				assignTypeHint(dst, srcTypeHint)
-			} else if srcTypeHint == "" {
-				assignTypeHint(src, dstTypeHint)
-			}
-		}
+// Two prims have compatible types if they have the same base type (in their
+// Value field) and the same data format kind, if any.
+func haveCompatibleTypes(dst, src *pb.Primitive) bool {
+	if dst == nil || src == nil {
+		return false
 	}
 
-	// Special case: If there are otherwise no conflicts, then merge data
-	// formats.  However, if there are conflicts, then leave data formats
-	// unmerged in their respective objects.
-	srcDataFormats := getDataFormats(src)
-	dstDataFormats := getDataFormats(dst)
-	assignDataFormats(src, nil)
-	assignDataFormats(dst, nil)
-
-	// Special case: If there are otherwise no conflicts, then merge example
-	// values. However, if there are conflicts, then leave example values
-	// unmerged in their respective objects.
-	srcExampleValues := src.ExampleValues
-	dstExampleValues := dst.ExampleValues
-	// Temporarily assign nil to example values for comparison.
-	src.ExampleValues = nil
-	dst.ExampleValues = nil
-
-	// Note: witnesses should not contain actual values and we've taken out data
-	// formats and example values above, so simple equality comparison works to
-	// determine whether 2 Data proto have conflict.
-	if !proto.Equal(dst, src) {
-		// Reinstate original data formats
-		assignDataFormats(src, srcDataFormats)
-		assignDataFormats(dst, dstDataFormats)
-
-		// Reinstate original example values
-		src.ExampleValues = srcExampleValues
-		dst.ExampleValues = dstExampleValues
+	cmpDst := &pb.Primitive{Value: dst.Value, FormatKind: dst.FormatKind}
+	cmpSrc := &pb.Primitive{Value: src.Value, FormatKind: src.FormatKind}
+	return proto.Equal(cmpDst, cmpSrc)
+}
```

Comment:
An unformatted int should subsume a formatted int, and should therefore be compatible. Similarly for other types, of course.

Reply:
Hm, true.  I don't think the existing code does that, though.  If you don't mind, I'll land this PR with just the refactoring and then do another to properly move up the type lattice during melding.

Reply:
I'll add some test cases to check this behavior first though.

Reply:
Added tests confirming that the existing behavior does not merge up the type lattice: https://github.com/akitasoftware/superstar/pull/636